### PR TITLE
Implement Simulation-Extractable ppzkSNARKs per [GM17]

### PR DESCRIPTION
**Pull request by popoffka**
This pull request implements the proof system presented in [Groth and Maller's IACR-CRYPTO-2017 paper](https://eprint.iacr.org/2017/540.pdf).

Unlike the system from [BCTV14a], which internally operates on Quadratic Arithmetic Programs (which arise from R1CSs, systems of constraints of the form <a, w> * <b, w> = <c, w>), [GM17] operates on Square Arithmetic Programs (which arise from constraints of the form (<a, w>)² = <c, w>). It turns out that an R1CS constraint can be converted into two equivalent SAP-compatible constraints at the cost of introducing an extra variable. This pull request includes an implementation of SAPs along with an R1CS→SAP reduction that follows the existing R1CS→QAP reduction but also performs this additional conversion.

It should be noted that the proof system relies on an assumption that is stronger than KEA used in most LIP SNARK works, but weaker than generic group model (used in Groth'16).

When compared to [BCTV14a] on the same input R1CS systems, [GM17] is slower but produces shorter proofs (just 2 elements in G1 and 1 in G2).

Here are the results of some benchmarking of our implementations of [BCTV14a] and [GM17] on an R1CS instance with 100 000 constraints and 1000 inputs:

Indicator | [BCTV14a] (`r1cs_ppzksnark`) | [GM17] (`r1cs_se_ppzksnark`) | delta
---|---|---|---|
Generator runtime | 16.91s | 59.95s | +254%
Prover runtime | 19.27s | 19.45s | +0.9%
Verifier runtime | 0.03s | 0.03s | 0
Proving key size | 255954523 bits | 322670816 bits | +26%
Verification key size | 322310 bits | 257292 bits | **-20%**
Proof size | 2294 bits | 1019 bits | **-56%**

### CODE REVIEW

File: libsnark/zk_proof_systems/ppzksnark/r1cs_se_ppzksnark/r1cs_se_ppzksnark.hpp [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+
+    size_t G1_size() const
+    {
+        return A_query.size() + C_query_1.size() + C_query_2.size() + 3
+               + G_gamma2_Z_t.size();
+    }
+
+    size_t G2_size() const
+    {
+        return B_query.size();
```
Comment (howardwu) [inline libsnark/zk_proof_systems/ppzksnark/r1cs_se_ppzksnark/r1cs_se_ppzksnark.hpp]: Should this be `return B_query.size() + 1` to account for `H_gamma_Z`?

Comment (popoffka) [inline libsnark/zk_proof_systems/ppzksnark/r1cs_se_ppzksnark/r1cs_se_ppzksnark.hpp]: You're right. Let me double-check all the other *_size() functions and commit a fix.

Comment (popoffka) [inline libsnark/zk_proof_systems/ppzksnark/r1cs_se_ppzksnark/r1cs_se_ppzksnark.hpp]: Fixed. Thanks for pointing this out, @howardwu .

Comment (howardwu) [inline libsnark/zk_proof_systems/ppzksnark/r1cs_se_ppzksnark/r1cs_se_ppzksnark.hpp]: Thanks for the quick fix! Can you update the proving key size and delta table entries in the PR description (if the change is noticeable)?

Comment (popoffka) [inline libsnark/zk_proof_systems/ppzksnark/r1cs_se_ppzksnark/r1cs_se_ppzksnark.hpp]: @howardwu done.

**Pull request merged**